File: 백준/Gold/17298. 오큰수/오큰수.js
const path = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
let input = require('fs').readFileSync(path).toString().trim().split(/\s+/).map(Number);

const [N, ...numbers] = input;

const stack = [];
const result = Array(N).fill(-1);

for (let i = 0; i < N; i++) {
  const currentNumber = numbers[i];
  // 아직 오큰수 못찾은 경우
  while (stack.length > 0 && numbers[stack[stack.length - 1]] < currentNumber) {
    // 오큰수 할당
    result[stack.pop()] = currentNumber;
  }
  stack.push(i);
}
console.log(result.join(' '));

File: 백준/Gold/17298. 오큰수/오큰수.py
import sys
input = sys.stdin.readline

N = int(input())
stack = []
result = [-1] * N

nums = list(map(int, input().split()))

for i in range(N):
  num = nums[i]
  while stack and nums[stack[-1]] < num:
    idx = stack.pop()
    result[idx] = num
  stack.append(i)

print(" ".join(map(str, result)))